src/openforms/submissions/cosigning.py

```diff
@@ -0,0 +1,195 @@
+
+logger = logging.getLogger(__name__)
+
+type CosignData = CosignV1Data | CosignV2Data
+
+
+class CosignV1Data(TypedDict):
+    """
+    The shape of data stored in
+    :attr:`openforms.submissions.models.Submission.co_sign_data`.
+    """
+
+    plugin: str
+    identifier: str
+    representation: NotRequired[str]
+    co_sign_auth_attribute: AuthAttribute
+    fields: JSONObject
+
+
+class CosignV2Data(FormAuth):
+    """
+    The shape of data stored in
+    :attr:`openforms.submissions.models.Submission.co_sign_data`.
+
+    In cosign v2, the data inherits from the standard :class:`FormAuth` data, so it
+    contains the auth plugin, attribute and identifier value.
+    """
+
+    cosign_date: NotRequired[str]
+    """
+    ISO-8601 formatted datetime indicating when the cosigner confirmed the submission.
+
+    The key may be absent in legacy submissions, which is why it's marked as
+    ``NotRequired``. It was added in Open Forms 2.7.0
+
+    .. todo:: DeprecationWarning -> remove NotRequired in Open Forms 4.0.
+    """
+
+
+class CosignState:
+    """
+    Encapsulate all the cosign state of a submission.
+    """
+
+    def __init__(self, submission: Submission):
+        self.submission = submission
+
+    def __repr__(self):
+        reference = (
+            self.submission.public_registration_reference or self.submission.uuid
+        )
+        return f"<CosignState submission=<Submission reference={reference}>>"
+
+    def _find_component(self) -> Component | None:
+        """
+        Discover the Formio cosign component in the submission/form.
+        """
+        # use the complete view on the Formio component tree(s)
+        configuration_wrapper = self.submission.total_configuration_wrapper
+
+        # there should only be one cosign component, so we check our flatmap to find the
+        # component.
+        for component in configuration_wrapper.component_map.values():
+            if component["type"] == "cosign":
+                return component
+        return None
+
+    @property
+    def is_required(self) -> bool:
+        """
+        Determine if cosigning is required for the submission or not.
+
+        Cosign is considered required in either of the following conditions:
+
+        * the cosign component in the form is required (even after evaluating form logic)
+        * the cosign component in the form is optional and an email address has been
+          provided - this indicates the submitter wanted someone to cosign it given the
+          choice.
+
+        .. note:: This does not consider cosign v1, as the cosigning is required to be
+           able to complete the submission.
+        """
+        cosign_component = self._find_component()
+        if cosign_component is None:
+            return False
+
+        # if the component itself is marked as required, we know for sure cosigning is
+        # required
+        required = cosign_component.get("validate", {}).get("required", False)
```

Comment:
for simplicity sake - glom is a shotgun and doesn't care about types - it will happily apply `.` syntax to objects and dicts and do magic stuff. This is much more strongly typed and clear what kind of data we're working with.

And while `glom` is quite a useful tool, it really is intended to process external data rather than being "just a convenience" or syntactic sugar.